REGRESSION (291444@main): UserMessageHandler.postMessage of unserializable types should result in WKScriptMessage.body of NSNull

achristensen07 opened the pull request:
#### 869c1b5dd5fb0db47a4fd9cc580e9d3f47900b5a
<pre>
REGRESSION (291444@main): UserMessageHandler.postMessage of unserializable types should result in WKScriptMessage.body of NSNull
<a href="https://bugs.webkit.org/show_bug.cgi?id=290646">https://bugs.webkit.org/show_bug.cgi?id=290646</a>
<a href="https://rdar.apple.com/146688485">rdar://146688485</a>

Reviewed by Abrar Rahman Protyasha and Sihui Liu.

When sending a non-serializable JS type from the web content process to become an ObjC type in the UI process,
it becomes WKErrorJavaScriptResultTypeIsUnsupported if it is the result of something with an error parameter
like the completion handler of WKWebView.evaluateJavaScript.  Before 291444@main if such a type were sent via
UserMessageHandler.postMessage it would arrive in the UI process as NSNull.null, but that change made it skip
the SerializedScriptValue filtering in the web content process and arrive as an NSDictionary, which broke
compatibility with some apps.  This restores the prior behavior and makes the JavaScriptEvaluationResult
constructor private to make it harder to forget the call to JavaScriptEvaluationResult::extract.

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::extract):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::extract):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
* Source/WebKit/UIProcess/UserContent/WebScriptMessageHandler.h:
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.cpp:
(WebKit::WebUserContentControllerProxy::didPostMessage):
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.h:
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.messages.in:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
(WebKit::WebInspectorUIExtensionController::evaluateScriptForExtension):
(WebKit::WebInspectorUIExtensionController::evaluateScriptInExtensionTab):
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
(TEST(WKWebView, EvaluateJavaScriptErrorCases)):

Canonical link: <a href="https://commits.webkit.org/292881@main">https://commits.webkit.org/292881@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fd7429e30338b7340f7d8d4ffcefe789d2d284e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16953 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102414 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47856 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17246 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25402 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74146 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13042 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88009 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54487 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5890 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47299 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82838 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104434 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24405 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83189 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24778 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84133 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82607 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27158 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4834 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15721 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24369 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24191 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27505 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25765 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -60,6 +60,11 @@ JavaScriptEvaluationResult::JavaScriptEvaluationResult(JSGlobalContextRef contex
+
+std::optional<JavaScriptEvaluationResult> JavaScriptEvaluationResult::extract(JSGlobalContextRef context, JSValueRef value)
+{
+    return JavaScriptEvaluationResult { context, value };
```
This never returns a nullopt... is `std::optional` necessary?

aprotyas replied:
Ignore, I kept reading and saw the other definition in the .mm file

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
```diff
@@ -146,14 +146,15 @@ - (void)removeAllContentRuleLists
             RetainPtr webView = page.cocoaView();
             if (!webView)
                 return;
+            RetainPtr<id> body = jsMessage ? jsMessage->toID() : NSNull.null;
```
nit:

```suggestion
            RetainPtr<id> body = jsMessage.transform([](auto result) {
                return result.toID();
            }).value_or(NSNull.null);
```

aprotyas replied:
also, you may be able to drop `<id>` because CTAD

achristensen07 replied:
I like dropping the template parameters.  I think the optional::transform is a bit verbose for such a simple case where a ternary operator works.

aprotyas commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -405,9 +405,9 @@ class ScriptMessageClientGtk final : public WebScriptMessageHandler::Client {
+    void didPostMessage(WebPageProxy&, FrameInfoData&&, API::ContentWorld&, std::optional<JavaScriptEvaluationResult>&& jsMessage) override
     {
-        Ref serializedScriptValue = API::SerializedScriptValue::createFromWireBytes(jsMessage.wireBytes());
+        Ref serializedScriptValue = API::SerializedScriptValue::createFromWireBytes(jsMessage->wireBytes());
```
is this optional guaranteed to be not empty?

achristensen07 replied:
Kind of because the non-Cocoa implementation of JavaScriptEvaluationResult::extract always returns non-nullopt.  That's very fragile though.  I'll add an early return in case that changes.

aprotyas commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -428,15 +428,15 @@ class ScriptMessageClientGtk final : public WebScriptMessageHandler::Client {
 
         WebKitScriptMessageReply* message = webKitScriptMessageReplyCreate(WTFMove(completionHandler));
-        Ref serializedScriptValue = API::SerializedScriptValue::createFromWireBytes(jsMessage.wireBytes());
+        Ref serializedScriptValue = API::SerializedScriptValue::createFromWireBytes(jsMessage->wireBytes());
```
Same question about emptiness of the optional value

achristensen07 committed:
REGRESSION (291444@main): UserMessageHandler.postMessage of unserializable types should result in WKScriptMessage.body of NSNull
https://bugs.webkit.org/show_bug.cgi?id=290646
rdar://146688485

Reviewed by Abrar Rahman Protyasha and Sihui Liu.

When sending a non-serializable JS type from the web content process to become an ObjC type in the UI process,
it becomes WKErrorJavaScriptResultTypeIsUnsupported if it is the result of something with an error parameter
like the completion handler of WKWebView.evaluateJavaScript.  Before 291444@main if such a type were sent via
UserMessageHandler.postMessage it would arrive in the UI process as NSNull.null, but that change made it skip
the SerializedScriptValue filtering in the web content process and arrive as an NSDictionary, which broke
compatibility with some apps.  This restores the prior behavior and makes the JavaScriptEvaluationResult
constructor private to make it harder to forget the call to JavaScriptEvaluationResult::extract.

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::extract):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::extract):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
* Source/WebKit/UIProcess/Inspector/socket/RemoteInspectorProtocolHandler.cpp:
* Source/WebKit/UIProcess/UserContent/WebScriptMessageHandler.h:
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.cpp:
(WebKit::WebUserContentControllerProxy::didPostMessage):
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.h:
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.messages.in:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
(WebKit::WebInspectorUIExtensionController::evaluateScriptForExtension):
(WebKit::WebInspectorUIExtensionController::evaluateScriptInExtensionTab):
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
(TEST(WKWebView, EvaluateJavaScriptErrorCases)):

Canonical link: https://commits.webkit.org/292881@main
